go/obscuronode/host/host.go

```diff
@@ -470,43 +488,86 @@ func (a *Node) handleStoreSecretTx(t *obscurocommon.L1RespondSecretTx) bool {
+		return
+	}
 
-			jsonAttestation, err := json.Marshal(att)
-			if err == nil {
-				nodecommon.LogWithID(a.shortID, "Received attestation request: %s", jsonAttestation)
-			} else {
-				nodecommon.LogWithID(a.shortID, "Received attestation request but it was unprintable.")
-			}
+	att, err := nodecommon.DecodeAttestation(scrtReqTx.Attestation)
+	if err != nil {
+		nodecommon.LogWithID(a.shortID, "Failed to decode attestation. %s", err)
+		return
+	}
 
-			secret, err := a.EnclaveClient.ShareSecret(att)
-			if err != nil {
-				nodecommon.LogWithID(a.shortID, "Secret request failed, no response will be published. %s", err)
-				continue
-			}
+	jsonAttestation, err := json.Marshal(att)
+	if err == nil {
+		nodecommon.LogWithID(a.shortID, "Received attestation request: %s", jsonAttestation)
+	} else {
+		nodecommon.LogWithID(a.shortID, "Received attestation request but it was unprintable.")
+	}
+
+	secret, err := a.EnclaveClient.ShareSecret(att)
+	if err != nil {
+		nodecommon.LogWithID(a.shortID, "Secret request failed, no response will be published. %s", err)
+		return
+	}
+
+	l1tx := &obscurocommon.L1RespondSecretTx{
+		Secret:      secret,
+		RequesterID: att.Owner,
+		AttesterID:  a.ID,
+		HostAddress: att.HostAddress,
+	}
+	// TODO review: l1tx.Sign(a.attestationPubKey) doesn't matter as the waitSecret will process a tx that was reverted
+	a.broadcastTx(a.mgmtContractLib.CreateRespondSecret(l1tx, a.ethWallet.GetNonceAndIncrement(), false))
+}
 
-			l1tx := &obscurocommon.L1RespondSecretTx{
-				Secret:      secret,
-				RequesterID: att.Owner,
-				AttesterID:  a.ID,
-				HostAddress: att.HostAddress,
+// Whenever we receive a new shared secret response transaction, we update our list of P2P peers, as another aggregator
+// may have joined the network.
+func (a *Node) processSharedSecretResponse(_ *obscurocommon.L1RespondSecretTx) error {
+	// We make a call to the L1 node to retrieve the new list of aggregators. An alternative would be to check that the
+	// transaction succeeded, and if so, extract the additional host address from the transaction arguments. But we
+	// believe this would be more brittle than just asking the L1 contract for its view of the current aggregators.
+	msg, err := a.mgmtContractLib.GetHostAddresses()
+	if err != nil {
+		return err
+	}
+	response, err := a.ethClient.CallContract(msg)
+	if err != nil {
+		return err
+	}
+	decodedResponse, err := a.mgmtContractLib.DecodeCallResponse(response)
+	if err != nil {
+		return err
+	}
+	hostAddresses := decodedResponse[0]
+
+	// We filter down the list of retrieved addresses.
+	var filteredHostAddresses []string //nolint:prealloc
+	for _, hostAddress := range hostAddresses {
+		// We exclude our own address.
+		if hostAddress == a.config.P2PAddress {
+			continue
+		}
+
+		// We exclude any duplicate host addresses.
+		isDup := false
+		for _, existingHostAddress := range filteredHostAddresses {
+			if hostAddress == existingHostAddress {
+				isDup = true
```

Comment:
Can't believe some of the stuff Go doesn't give you out of the box, super gross haha. You can `break` here I think

Reply:
Nice.